Libraries/LibWeb/Layout/InlineLevelIterator.cpp

```diff
@@ -498,15 +499,26 @@ Optional<InlineLevelIterator::Item> InlineLevelIterator::next_without_lookahead(
         else
             m_text_node_context->is_first_chunk = false;
 
+        if (!m_text_node_context->chunk_iterator.peek(0).has_value())
+            m_text_node_context->is_last_chunk = true;
+
         auto chunk_opt = m_text_node_context->chunk_iterator.next();
+        auto is_empty_editable = false;
         if (!chunk_opt.has_value()) {
-            m_text_node_context = {};
-            skip_to_next();
-            return next_without_lookahead();
-        }
+            auto const is_only_chunk = m_text_node_context->is_first_chunk && m_text_node_context->is_last_chunk;
+            if (is_only_chunk && text_node.text_for_rendering().is_empty()) {
+                if (auto const* shadow_root = as_if<DOM::ShadowRoot>(text_node.dom_node().root()))
+                    is_empty_editable = shadow_root->host() && (is<HTML::FormAssociatedTextControlElement>(shadow_root->host()) || shadow_root->host()->is_editable_or_editing_host());
```

Comment:
If you have a `FormAssociatedTextControlElement`, shouldn't we also look at `::is_mutable()`? If the control element is disabled or readonly, we probably shouldn't entertain adding empty fragments.